floors/enforce.go

```diff
@@ -0,0 +1,202 @@
+package floors
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/exchange/entities"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// EnforceFloors does floors enforcement for bids from all bidders based on floors provided in request, account level floors config
+func EnforceFloors(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, account config.Account, conversions currency.Conversions) (map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, []error, []*entities.PbsOrtbSeatBid) {
+	rejectionErrs := []error{}
+	rejectedBids := []*entities.PbsOrtbSeatBid{}
+
+	requestExt, err := bidRequestWrapper.GetRequestExt()
+	if err != nil {
+		return seatBids, []error{errors.New("Error in getting request extension")}, rejectedBids
+	}
+
+	if isPriceFloorsDisabled(account, bidRequestWrapper) {
+		return seatBids, []error{errors.New("Floors feature is disabled at account or in the request")}, rejectedBids
+	}
+
+	if !isFloorsSignallingSkipped(requestExt) && isValidImpBidFloorPresent(bidRequestWrapper.BidRequest) {
+		if enforceFloors := isSatisfiedByEnforceRate(requestExt, account.PriceFloors.EnforceFloorsRate, rand.Intn); enforceFloors {
+			enforceDealFloors := account.PriceFloors.EnforceDealFloors && getEnforceDealsFlag(requestExt)
+
+			seatBids, rejectionErrs, rejectedBids = enforceFloorToBids(bidRequestWrapper, seatBids, conversions, enforceDealFloors)
+		}
+	}
+	return seatBids, rejectionErrs, rejectedBids
+}
+
+// enforceFloorToBids function does floors enforcement for each bid,
+// The bids returned by each partner below bid floor price are rejected and remaining eligible bids are considered for further processing
+func enforceFloorToBids(bidRequestWrapper *openrtb_ext.RequestWrapper, seatBids map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, conversions currency.Conversions, enforceDealFloors bool) (map[openrtb_ext.BidderName]*entities.PbsOrtbSeatBid, []error, []*entities.PbsOrtbSeatBid) {
+	errs := []error{}
+	rejectedBids := []*entities.PbsOrtbSeatBid{}
+	impMap := make(map[string]*openrtb_ext.ImpWrapper, bidRequestWrapper.LenImp())
+
+	for _, v := range bidRequestWrapper.GetImp() {
+		impMap[v.ID] = v
+	}
+
+	for bidderName, seatBid := range seatBids {
+		eligibleBids := make([]*entities.PbsOrtbBid, 0, len(seatBid.Bids))
+		for _, bid := range seatBid.Bids {
+
+			reqImp, ok := impMap[bid.Bid.ImpID]
+			if !ok {
+				continue
+			}
+
+			reqImpCur := reqImp.BidFloorCur
+			if reqImpCur == "" {
+				reqImpCur = defaultCurrency
+				if bidRequestWrapper.Cur != nil {
+					reqImpCur = bidRequestWrapper.Cur[0]
+				}
+			}
+			updateBidExtWithFloors(reqImp, bid, reqImpCur)
+
+			if !isPriceFloorsEnforcementDisabled(bidRequestWrapper) {
+				dealBid := checkDealBidForEnforcement(bid, enforceDealFloors)
```

Comment:
Nitpick: thoughts on replacing `checkDealBidForEnforcement` with a function like `hasDealID()` and instead writing `if hasDealID(bid) && !enforceDealFloors {` to make this more readable and avoid deferring on the evaluation of `enforceDealFloors` to the helper function?

Reply:
Done